File: datasets/C++/Chatgpt/DataStructures/hard5.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

const int MAXN = 75005;
vector<int> tree[MAXN];
int dist[MAXN];

void bfs(int start, int n) {
    fill(dist, dist + n + 1, -1);
    queue<int> q;
    q.push(start);
    dist[start] = 0;

    while (!q.empty()) {
        int node = q.front(); q.pop();
        for (int v : tree[node]) {
            if (dist[v] == -1) {
                dist[v] = dist[node] + 1;
                q.push(v);
            }
        }
    }
}

int main() {
    int n, m;
    cin >> n >> m;

    for (int i = 0; i < n - 1; i++) {
        int u, v;
        cin >> u >> v;
        tree[u].push_back(v);
        tree[v].push_back(u);
    }

    // First BFS to find a far node
    bfs(1, n);
    int far1 = 1;
    for (int i = 1; i <= n; i++) {
        if (dist[i] > dist[far1])
            far1 = i;
    }

    bfs(far1, n);
    vector<pair<int, int>> points(m);
    for (int i = 0; i < m; i++) {
        cin >> points[i].first >> points[i].second;
    }

    auto tree_distance = [&](int u, int v) {
        return abs(dist[u] - dist[v]);
    };

    auto squared_distance = [&](pair<int, int> a, pair<int, int> b) {
        return tree_distance(a.first, b.first) + tree_distance(a.second, b.second);
    };

    int idx = 0, max_d = 0;
    for (int i = 0; i < m; i++) {
        int d = squared_distance(points[0], points[i]);
        if (d > max_d) {
            max_d = d;
            idx = i;
        }
    }

    for (int i = 0; i < m; i++) {
        int d = squared_distance(points[idx], points[i]);
        max_d = max(max_d, d);
    }

    cout << max_d << endl;
    return 0;
}

File: datasets/Python/Chatgpt/DataStructures/hard5.py
from collections import deque, defaultdict

def bfs(start, tree, n):
    dist = [-1] * (n + 1)
    dist[start] = 0
    q = deque([start])
    while q:
        node = q.popleft()
        for nei in tree[node]:
            if dist[nei] == -1:
                dist[nei] = dist[node] + 1
                q.append(nei)
    return dist

def main():
    n, m = map(int, input().split())
    tree = defaultdict(list)
    
    for _ in range(n - 1):
        u, v = map(int, input().split())
        tree[u].append(v)
        tree[v].append(u)

    # One BFS from any point to find furthest node
    def get_furthest(node):
        dist = bfs(node, tree, n)
        max_dist = -1
        far_node = -1
        for i in range(1, n + 1):
            if dist[i] > max_dist:
                max_dist = dist[i]
                far_node = i
        return far_node, dist

    # First BFS to find farthest node from 1
    far1, _ = get_furthest(1)
    # Second BFS from far1 to get all distances
    _, dist_from_far1 = get_furthest(far1)

    points = []
    for _ in range(m):
        x, y = map(int, input().split())
        points.append((x, y))

    def tree_distance(u, v):
        return abs(dist_from_far1[u] - dist_from_far1[v])

    def squared_distance(p1, p2):
        return tree_distance(p1[0], p2[0]) + tree_distance(p1[1], p2[1])

    max_dist = 0
    idx = 0
    for i in range(m):
        d = squared_distance(points[0], points[i])
        if d > max_dist:
            max_dist = d
            idx = i

    for i in range(m):
        d = squared_distance(points[idx], points[i])
        if d > max_dist:
            max_dist = d

    print(max_dist)

main()